impala/hiveserver2.py

```diff
@@ -949,15 +983,32 @@ def _convert_values(self, type_, is_null, values):
         return values
 
     def __len__(self):
-        return len(self.columns[0]) if len(self.columns) > 0 else 0
+        return self.remaining_rows
 
     def pop(self):
+        self.remaining_rows -= 1
         return tuple([c.pop() for c in self.columns])
 
     def __str__(self):
         col_string = ','.join([str(col) for col in self.columns])
         return 'CBatch({0})'.format(col_string)
 
+    def pop_many(self, row_count):
+        """Returns a list of tuples with min('row_count', rows in batch) elements."""
+        row_count = min(row_count, self.remaining_rows)
+        self.remaining_rows -= row_count
+        col_count = len(self.columns)
+        # 'dataset' holds all rows x columns in list in row major order.
+        # The transposition of columnar data is done by writing 'dataset' per-column
+        # and then returning it per-row.
+        dataset = [None] * (col_count * row_count)
+        for col_id, col in enumerate(self.columns):
+            col.pop_to_preallocated_list(
```

Comment:
The return value of 'pop_to_preallocated_list' should always be equal to 'row_count', right? If so, please add an assert after this call that verifies that the post-condition is true.

Reply:
Done